Dockerfile

```diff
@@ -1,35 +1,87 @@
+# from the master branch of https://github.com/docker/docker.github.io
+#
+# Here is the sequence:
+# 1.  Set up the build
+# 2.  Fetch upstream resources
+# 3.  Build master
+# 4.  Copy static HTML from already-built archive images
+# 5.  Copy Nginx config
+#
+# When the image is run, it starts Nginx and serves the docs at port 4000
 
-# docs-base contains: GitHub Pages, nginx, wget, svn, and the docs archives,
-# running on Alpine. See the contents of docs-base at:
-# https://github.com/docker/docker.github.io/tree/docs-base
+# Get basic configs and Jekyll env
+FROM docs/docker.github.io:docs-builder AS builder
 
-# First, build non-edge (all of this is duplicated later -- that is on purpose)
+# Set the target again
+ENV TARGET=/usr/share/nginx/html
 
-# Copy master into target directory (skipping files / folders in .dockerignore)
-# These files represent the current docs
-COPY . md_source
+# Set the source directory to md_source
+ENV SOURCE=md_source
 
-# Move built html into md_source directory so we can reuse the target directory
-# to hold the static output.
-# Pull reference docs from upstream locations, then build the master docs
-# into static HTML in the "target" directory using Jekyll
-# then nuke the md_source directory.
+# Get the current docs from the checked out branch
+# ${SOURCE} will contain a directory for each archive
+COPY . ${SOURCE}
 
+####### START UPSTREAM RESOURCES ########
+# Set vars used by fetch-upstream-resources.sh script
 ## Branch to pull from, per ref doc
 ## To get master from svn the svn branch needs to be 'trunk'. To get a branch from svn it needs to be 'branches/branchname'
 
 # Engine
-ENV ENGINE_SVN_BRANCH="branches/17.06.x"
-ENV ENGINE_BRANCH="17.06.x"
+ENV ENGINE_SVN_BRANCH="branches/17.09.x"
+ENV ENGINE_BRANCH="17.09.x"
 
 # Distribution
 ENV DISTRIBUTION_SVN_BRANCH="branches/release/2.6"
 ENV DISTRIBUTION_BRANCH="release/2.6"
 
-RUN /sbin/apk --update add bash \
-  && bash ./md_source/_scripts/fetch-upstream-resources.sh \
-	&& jekyll build -s md_source -d target --config md_source/_config.yml \
-	&& rm -rf target/apidocs/layouts \
-	&& find target -type f -name '*.html' -print0 | xargs -0 sed -i 's#href="https://docs.docker.com/#href="/#g' \
-	&& rm -rf md_source
+# Fetch upstream resources
+RUN bash ./${SOURCE}/_scripts/fetch-upstream-resources.sh ${SOURCE}
+####### END UPSTREAM RESOURCES ########
+
+
+# Build the static HTML, now that everything is in place
+
+RUN jekyll build -s ${SOURCE} -d ${TARGET} --config ${SOURCE}/_config.yml
+
+# Fix up some links, don't touch the archives
+RUN find ${TARGET} -type f -name '*.html' | grep -vE "v[0-9]+\." | while read i; do sed -i 's#href="https://docs.docker.com/#href="/#g' "$i"; done
+
+# BUILD OF MASTER DOCS IS NOW DONE!
+# Reset to alpine so we don't get any docs source or extra apps
+FROM nginx:alpine
+
+# Set the target again
+ENV TARGET=/usr/share/nginx/html
+
+# Get the built docs output from the previous step
+COPY --from=builder ${TARGET} ${TARGET}
```

Comment:
We should move this one after the archives, so that the archives copy step can be cached, and a change in master doesn't require updating the archives.